x/ccv/provider/keeper/punish_validators_test.go

```diff
@@ -0,0 +1,136 @@
+package keeper_test
+
+import (
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	evidencetypes "github.com/cosmos/cosmos-sdk/x/evidence/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	cryptotestutil "github.com/cosmos/interchain-security/v2/testutil/crypto"
+	testkeeper "github.com/cosmos/interchain-security/v2/testutil/keeper"
+	"github.com/cosmos/interchain-security/v2/x/ccv/provider/types"
+	"github.com/golang/mock/gomock"
+)
+
+func TestJailAndTombstoneValidator(t *testing.T) {
+	providerConsAddr := cryptotestutil.NewCryptoIdentityFromIntSeed(7842334).ProviderConsAddress()
+	testCases := []struct {
```

Comment:
If my understanding is correct that those tests check the order of operations (i.e., method calls) taking place in `JailAndTombstoneValidator`. So, the first test checks that `GetValidatorByConsAddr` is called. Then, the third test checks that if `GetValidatorByConsAddr` passes and since we are not unbonded, then we call `IsTombstoned`, etc.

Maybe it would make sense for `JailAndTombstoneValidator` to return an `err`or on whether it succeeds or not and then test depending (on the mocks we have) on whether `JailAndTombstoneValidator` succeeds? Or is this not possible because the underlying `JailUntil and `Tombstone` from the `slashing` module do not return an `err`or?

Reply:
The success is tested by the number of expected mocks per test case.
For instance, when the validator is unbonded, the test expects to have one unique call to 'GetValidatorByConsAddr` or would fail.